libgnucash/app-utils/gnc-option.hpp

```diff
@@ -0,0 +1,254 @@
+ *  @brief Represents the public interface for an option.
+ *  Polymorphism is provided by a std::variant member containing GncOptionValue
+ *  types.
+*/
+
+class GncOption
+{
+public:
+    template <typename OptionType,
+              typename std::enable_if_t<is_OptionClassifier_v<OptionType>,
+                               int>  = 0>
+
+    GncOption(OptionType option) :
+        m_option{std::make_unique<GncOptionVariant>(option)} {}
+    template <typename ValueType,
+              typename std::enable_if_t<!is_OptionClassifier_v<ValueType>,
+                               int>  = 0>
+    GncOption(const char* section, const char* name,
+              const char* key, const char* doc_string,
+              ValueType value,
+              GncOptionUIType ui_type = GncOptionUIType::INTERNAL);
+    template <typename ValueType> void set_value(ValueType value);
+    template <typename ValueType> void set_default_value(ValueType value);
+    template <typename ValueType> ValueType get_default_value() const;
+    template <typename ValueType> ValueType get_value() const;
+    void reset_default_value();
+
+    const std::string& get_section() const;
+    const std::string& get_name() const;
+    const std::string& get_key() const;
+    const std::string& get_docstring() const;
+    void set_ui_item(GncOptionUIItemPtr&& ui_elem);
+    const GncOptionUIType get_ui_type() const;
+    void set_ui_item_selectable(bool) const noexcept;
+    GncOptionUIItem* const get_ui_item() const;
+    void set_ui_item_from_option();
+    void set_option_from_ui_item();
+    void make_internal();
+    bool is_changed() const noexcept;
+/** @returns false unless m_option contains a GncOptionMultiselectValue or
+ * GncOptionAccountListValue for which multiple selections have been enabled.
+ */
+    bool is_multiselect() const noexcept;
+/** Implemented only for GncOptionNumericRange */
+    template <typename ValueType> void get_limits(ValueType&, ValueType&,
+                                                  ValueType&) const noexcept;
+/** Not implemented for GncOptionValue. */
+    template <typename ValueType> bool validate(ValueType value) const;
+/** Implemented only for GncOptionMultiselectValue. */
+    std::size_t num_permissible_values() const;
+/** Implemented only for GncOptionMultiselectValue. */
+    std::size_t permissible_value_index(const char* value) const;
+/** Implemented only for GncOptionMultiselectValue. */
+    const char* permissible_value(std::size_t index) const;
+/** Implemented only for GncOptionMultiselectValue. */
+    const char* permissible_value_name(std::size_t index) const;
+/** Implemented only for GncOptionAccountListValue. */
+    GList* account_type_list() const noexcept;
+    bool is_alternate() const noexcept;
+    void set_alternate(bool) noexcept;
+    std::ostream& out_stream(std::ostream& oss) const;
+    std::istream& in_stream(std::istream& iss);
+
+    friend GncOptionVariant& swig_get_option(GncOption*);
```

Comment:
I presume that's in here because swig needs access to the private m_option member ?
I see you have (re)implemented this function in gnc-optiondb.i.

Reply:
Right. More specifically there are legacy guile functions whose implementations need API that I don't want to expose so it's buried in gnc_optiondb.i.

Reply:
Ok.

Reply:
This particular `friend` declaration brings me to swig's [`%extend`](http://www.swig.org/Doc4.0/SWIGDocumentation.html#SWIGPlus_class_extension) macro which might be a way to avoid this friend declaration. Though probably not exactly with your current implementation. The limitation of %exend is that it can only access public class members while you probably are using private ones. But I bring it up anyway. It may be conductive to another solution that doesn't require friend hoops just for the bindings.